### PR TITLE
Bug 1144784 - [System] Asking for more than one permission in a short period of time is broken r=alive,gasolin

### PR DESCRIPTION
**Pull request by borjasalguero**

### CODE REVIEW

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

File: apps/system/js/permission_manager.js [hunk 1]
```diff
@@ -10,6 +10,9 @@
   function PermissionManager() {
   }
 
+
+
+
```
Comment (alivedise) [inline apps/system/js/permission_manager.js]: nit: remove this empty line

File: apps/system/js/permission_manager.js [hunk 2]
```diff
@@ -102,6 +104,38 @@
...
     },
 
+    getStrings: function getStrings(detail) {
+      var permissionID = 'perm-' + this.permissionType.replace(':', '-');
+      var app = applications.getByManifestURL(detail.manifestURL);
+      var _ = navigator.mozL10n.get;
+
+      var message = '';
+      if (detail.isApp) {
+        var name = new ManifestHelper(app.manifest).name;
```
Comment (alivedise) [inline apps/system/js/permission_manager.js]: Where does this being used?

File: apps/system/js/permission_manager.js [hunk 3]
```diff
@@ -171,6 +206,15 @@
       }
       this.buttons.dataset.items = 2;
       this.no.style.display = 'inline';
+
```
Comment (alivedise) [inline apps/system/js/permission_manager.js]: nit: remove this empty line

File: apps/system/js/permission_manager.js [hunk 4]
```diff
@@ -339,6 +326,75 @@
...
+          if (detail.permissions.hasOwnProperty(permission2)) {
+            // gecko might not support audio/video option
+            if (detail.permissions[permission2].length > 0) {
+              this.currentChoices[permission2] =
+                detail.permissions[permission2][0];
+            }
+          }
+        }
+      }
+      // ****** HASTA AQUI LO QUE ESTABA EN EL handleEvent****
```
Comment (gasolin) [inline apps/system/js/permission_manager.js]: what's that?

File: apps/system/js/permission_manager.js [hunk 5]
```diff
@@ -102,6 +104,38 @@
       });
     },
 
+    getStrings: function getStrings(detail) {
```
Comment (gasolin) [inline apps/system/js/permission_manager.js]: please put this function after `stop` function

File: shared/locales/permissions/permissions.en-US.properties [hunk 6]
```diff
@@ -27,6 +27,8 @@ perm-browser=Inspect Web Pages
 perm-camera=Cameras
 perm-camera-selector-appRequest=Select a camera for {{app}}.
 perm-camera-selector-webRequest=Select a camera for {{site}}.
+perm-camera-appRequest={{app}} would like to use your camera.
```
Comment (gasolin) [inline shared/locales/permissions/permissions.en-US.properties]: If I remember correctly, it's not need to have `perm-camera-app/webRequest`. Do you find a scenario that show such wording?

File: apps/system/js/permission_manager.js [hunk 7]
```diff
@@ -171,6 +206,15 @@
       }
       this.buttons.dataset.items = 2;
       this.no.style.display = 'inline';
+
+    },
+
+    /**
+     * Queue or show the permission prompt
```
Comment (gasolin) [inline apps/system/js/permission_manager.js]: the content seems only queue the permission prompt

File: apps/system/js/permission_manager.js [hunk 8]
```diff
@@ -182,70 +226,10 @@
...
-                this.isVideo = true;
-
-                LazyLoader.load('shared/js/template.js');
-              }
-              if ('audio-capture' === detail.permission) {
-                this.isAudio = true;
-              }
-            }
+          if (!!this.currentRequestId) {
+            this.queuePrompt(detail);
```
Comment (gasolin) [inline apps/system/js/permission_manager.js]: Nice :+1:

File: shared/locales/permissions/permissions.en-US.properties (hunk 6 above)
Comment (borjasalguero) [inline shared/locales/permissions/permissions.en-US.properties]: After some discussion, 'camera' seems to be a good candidate to be one of the API's that could be opened to the web (actually you can take a photo with WebRTC right now), so I've added it for having this scenario ready for the near future (actually I've done some tests about it, and seems to be feasible)

File: apps/system/js/permission_manager.js (hunk 4 above)
Comment (borjasalguero) [inline apps/system/js/permission_manager.js]: Sorry for this! Removing this comment in Spanish :)

File: apps/system/js/permission_manager.js (hunk 3 above)
Comment (borjasalguero) [inline apps/system/js/permission_manager.js]: +1

File: apps/system/js/permission_manager.js (hunk 2 above)
Comment (borjasalguero) [inline apps/system/js/permission_manager.js]: Dead code. Removing

File: apps/system/js/permission_manager.js (hunk 1 above)
Comment (borjasalguero) [inline apps/system/js/permission_manager.js]: Removing!

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by try-server-hook** [comment]
[borjasalguero Borja Salguero (borjasalguero)](https://github.com/borjasalguero) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=30ec296630fd)

File: shared/locales/permissions/permissions.en-US.properties (hunk 6 above)
Comment (borjasalguero) [inline shared/locales/permissions/permissions.en-US.properties]: Furthermore, 'camera' is privileged, so at least 'app' is a must (you could have a new 'camera' app from the marketplace)

File: apps/system/test/unit/permission_manager_test.js [hunk 9]
```diff
@@ -20,16 +21,22 @@ function sendChromeEvent(evt_type, evt_permission, remember) {
   window.dispatchEvent(evt);
 }
 
+function createMediaEvent(evt_type, evt_permissions, isApp, isGranted) {
+  return {
+            'type': evt_type,
```
Comment (gasolin) [inline apps/system/test/unit/permission_manager_test.js]: nit: the intent would be 2 space right from `return`

File: apps/system/test/unit/permission_manager_test.js [hunk 10]
```diff
@@ -93,13 +102,35 @@ suite('system/permission manager', function() {
...
 
     test('permission-prompt', function() {
       assert.equal(permissionManager.overlay.dataset.type, 'test');
       assert.isTrue(permissionManager.handlePermissionPrompt.called);
+      assert.isTrue(permissionManager.queuePrompt.called);
+    });
+
+    test('permission-prompt queue', function() {
+      assert.equal(permissionManager.overlay.dataset.type, 'test');
+      assert.isTrue(permissionManager.handlePermissionPrompt.called);
```
Comment (gasolin) [inline apps/system/test/unit/permission_manager_test.js]: does above 2 lines duplicate from the above test?

File: apps/system/test/unit/permission_manager_test.js [hunk 11]
```diff
@@ -628,24 +793,44 @@ suite('system/permission manager', function() {
...
       });
 
-      test('should More info... link be hidden when we accept/deny the prompt', 
+      test('should "More info..." be hidden when we accept/deny the prompt',
         function() {
-          var yescallback = this.sinon.stub();
-          var nocallback = this.sinon.stub();
-          permissionManager.showPermissionPrompt(1, '', 'testMoreInfoText', 
-            yescallback, nocallback);
+          this.sinon.stub(
```
Comment (gasolin) [inline apps/system/test/unit/permission_manager_test.js]: put `this.sinon.stub(permissionManager, 'getStrings').returns({` in a line

File: apps/system/test/unit/permission_manager_test.js [hunk 12]
```diff
@@ -365,10 +391,9 @@ suite('system/permission manager', function() {
       window.dispatchEvent(evt);
     });
 
+
```
Comment (gasolin) [inline apps/system/test/unit/permission_manager_test.js]: remove this redundent blank line

File: apps/system/test/unit/permission_manager_test.js [hunk 13]
```diff
@@ -407,21 +481,73 @@ suite('system/permission manager', function() {
...
+      'manifestURL': 'app://uitest.gaiamobile.org/manifest.webapp',
+      'origin': 'app://uitest.gaiamobile.org/',
+      'manifest': {
+        'name': 'UITest',
+        'role': 'UITest',
+      },
+      target: {}
+    };
+
+
```
Comment (gasolin) [inline apps/system/test/unit/permission_manager_test.js]: please leave only one blank line between methods

File: apps/system/test/unit/permission_manager_test.js [hunk 14]
```diff
@@ -407,21 +481,73 @@ suite('system/permission manager', function() {
...
     test('permission-prompt', function() {
       assert.equal(permissionManager.permissionType, 'video-capture');
     });
 
-    test('permission id matched', function() {
-      assert.isTrue(permissionManager.requestPermission
-        .calledWithMatch('perm1', 'test', 'video-capture',
-        sinon.match.string, 'perm-video-capture-more-info'));
+
+    test('Web: All strings are matching', function() {
```
Comment (gasolin) [inline apps/system/test/unit/permission_manager_test.js]: :+1:

File: apps/system/test/unit/permission_manager_test.js [hunk 15]
```diff
@@ -628,24 +793,44 @@ suite('system/permission manager', function() {
...
+              moreInfoText: 'moreInfoText'
+            }
+          );
+
+          // Launch
+          sendMediaEvent('permission-prompt', {'audio-capture': ['']});
           assert.isFalse(
             permissionManager.moreInfo.classList.contains('hidden'));
+
+
```
Comment (gasolin) [inline apps/system/test/unit/permission_manager_test.js]: please only leave single blank line

File: apps/system/test/unit/permission_manager_test.js [hunk 16]
```diff
@@ -628,24 +793,44 @@ suite('system/permission manager', function() {
...
-            yescallback, nocallback);
-          assert.isTrue(
-            permissionManager.moreInfo.classList.contains('hidden'));
       });
-  });
+    });
+
+    test('"More info..." must not be shown if there is not text',
+      function() {
+        this.sinon.stub(
```
Comment (gasolin) [inline apps/system/test/unit/permission_manager_test.js]: put `this.sinon.stub(permissionManager, 'getStrings').returns({` in a line

File: apps/system/test/unit/permission_manager_test.js (hunk 9 above)
Comment (borjasalguero) [inline apps/system/test/unit/permission_manager_test.js]: Fixed!

File: apps/system/test/unit/permission_manager_test.js (hunk 10 above)
Comment (borjasalguero) [inline apps/system/test/unit/permission_manager_test.js]: +1. I will merge both tests into one.

Comment (borjasalguero) [inline apps/system/test/unit/permission_manager_test.js]: Actually I've added a new one in order to double check that `dispatchEvent` is called the right number of times :)

File: apps/system/test/unit/permission_manager_test.js (hunk 11 above)
Comment (borjasalguero) [inline apps/system/test/unit/permission_manager_test.js]: Fixing.

File: apps/system/test/unit/permission_manager_test.js (hunk 12 above)
Comment (borjasalguero) [inline apps/system/test/unit/permission_manager_test.js]: +1

File: apps/system/test/unit/permission_manager_test.js (hunk 13 above)
Comment (borjasalguero) [inline apps/system/test/unit/permission_manager_test.js]: +1

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=a9a8a17f29d6f7050c6c5ce052eb6910d5be7be9

**Pull request merged**